### PR TITLE
Refactor PConfiguration (Simplify Config Part 4)

### PR DESCRIPTION
**Pull request by nupplaphil**
FollowUp and similar to #7372 
- Move Adapter business logic to Configuration
- Remove all Adapter
- Bugfixing Preload PConfiguration again (instead of #6688 )

- Tested on local node
- Tested with JIT (didn't retrieve all system config values of the DB)
- Tested with Preload (retrieved all system config values of the DB)
- added tests to ensure that all config values will get loaded at once with preload configuration

And of course it already contains the fix of #7384 , so hopefully no sideeffects ...

This is the last of the refactorings. So I dropped the whole adapter directory.

### CODE REVIEW

File: src/Core/Config/PreloadPConfiguration.php [hunk 1]
```diff
@@ -0,0 +1,116 @@
+<?php
+
+namespace Friendica\Core\Config;
+
+use Friendica\Model;
+
+/**
+ * This class implements the preload Time configuration, which will cache
```
Comment (MrPetovan) [inline src/Core/Config/PreloadPConfiguration.php]: `-Time`

**Comment by annando** [comment]
Had you seen this?
```
[15-Jul-2019 20:13:30 Europe/Berlin] PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Friendica\Core\Config\Cache\PConfigCache::get() must be of the type integer, null given, called in /src/Core/Config/PConfiguration.php on line 87 and defined in /src/Core/Config/Cache/PConfigCache.php:65
Stack trace:
#0 /src/Core/Config/PConfiguration.php(87): Friendica\Core\Config\Cache\PConfigCache->get(NULL, 'feature', 'categories')
#1 /src/Core/PConfig.php(62): Friendica\Core\Config\PConfiguration->get(NULL, 'feature', 'categories', false, false)
#2 /src/Content/Feature.php(28): Friendica\Core\PConfig::get(NULL, 'feature', 'categories', false)
#3 /src/Content/Widget.php(300): Friendica\Content\Feature::isEnabled(NULL, 'categories')
#4 /src/Module/Profile.php(198): Friendica\Content\Widget::categories('https://pirati....', '')
#5 [internal in /src/Core/Config/Cache/PConfigCache.php on line 65
```

File: src/Core/Config/PreloadPConfiguration.php [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+		$this->config_loaded[$uid] = true;
+
+		// load the whole category out of the DB into the cache
+		$this->configCache->load($uid, $config);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	public function get(int $uid, string $cat, string $key, $default_value = null, bool $refresh = false)
```
Comment (MrPetovan) [inline src/Core/Config/PreloadPConfiguration.php]: The `$refresh` parameter is neither necessary nor supported with the preload configuration. Any `set`/`remove` operation will update the cache anyway, making the new value immediately available to further `get` calls. I'm guessing this is an artifact from the legacy JIT configuration.

File: src/Model/Config/PConfig.php [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+		$return = [];
+
+		if (empty($cat)) {
+			$configs = $this->dba->select('pconfig', ['cat', 'v', 'k'], ['uid' => $uid]);
+		} else {
+			$configs = $this->dba->select('üconfig', ['cat', 'v', 'k'], ['cat' => $cat, 'uid' => $uid]);
+		}
+
+		while ($config = $this->dba->fetch($configs)) {
+
```
Comment (MrPetovan) [inline src/Model/Config/PConfig.php]: Code standards: Please remove this superfluous new line.

File: src/Core/Config/PreloadPConfiguration.php (hunk 2 above)
Comment (nupplaphil) [reply src/Core/Config/PreloadPConfiguration.php]: Are you sure? I do find the `$refresh` parameter at the very first implementation of the preload config. 
https://github.com/friendica/friendica/blob/2a760bb770c015703567caa91054bcb3c9422fc0/src/Core/Config/PreloadConfigAdapter.php#L44-L56

https://github.com/friendica/friendica/blob/2a760bb770c015703567caa91054bcb3c9422fc0/src/Core/Config/PreloadPConfigAdapter.php#L48-L66

Comment (nupplaphil) [reply src/Core/Config/PreloadPConfiguration.php]: And what about config values which are changed after the initial load, which should be "forced" to refresh? I think that's the same reason for this parameter as for JIT

Comment (MrPetovan) [reply src/Core/Config/PreloadPConfiguration.php]: Ah, never mind, I forgot we have a daemon or a long-running worker. Both probably need the refresh feature even though I have no idea if they are using it (probably not).

**Comment by nupplaphil** [comment]
Seems like type hint now shows some issues ... 

> 
> 
> Had you seen this?
> 
> ```
> [15-Jul-2019 20:13:30 Europe/Berlin] PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Friendica\Core\Config\Cache\PConfigCache::get() must be of the type integer, null given, called in /src/Core/Config/PConfiguration.php on line 87 and defined in /src/Core/Config/Cache/PConfigCache.php:65
> Stack trace:
> #0 /src/Core/Config/PConfiguration.php(87): Friendica\Core\Config\Cache\PConfigCache->get(NULL, 'feature', 'categories')
> #1 /src/Core/PConfig.php(62): Friendica\Core\Config\PConfiguration->get(NULL, 'feature', 'categories', false, false)
> #2 /src/Content/Feature.php(28): Friendica\Core\PConfig::get(NULL, 'feature', 'categories', false)
> #3 /src/Content/Widget.php(300): Friendica\Content\Feature::isEnabled(NULL, 'categories')
> #4 /src/Module/Profile.php(198): Friendica\Content\Widget::categories('https://pirati....', '')
> #5 [internal in /src/Core/Config/Cache/PConfigCache.php on line 65
> ```

Seems like there are some issues when calling PConfig with an empty `$uid` .. I've to rethink about the type-hint, because there are a lot of `PConfig::get(local_user(), ...)` calls, which will return `false` in some circumstands too ...

> 
> 
> Do you plan on refactoring the node-wide configuration the same way as well or did I miss it?

see #7372

**Comment by MrPetovan** [comment]
Are you going to address this in this PR?

File: src/Core/Config/PreloadPConfiguration.php (hunk 2 above)
Comment (nupplaphil) [reply src/Core/Config/PreloadPConfiguration.php]: I found at least one occurence here:
https://github.com/friendica/friendica/blob/ff99a62584d256bef45a4053e2f6d505faf45c07/src/Util/ExAuth.php#L240

**Comment by nupplaphil** [comment]
It depends .. If I "just" remove the typehints, I can address it here too, but I'm not sure if this is the clean way .. What do you say?

**Comment by MrPetovan** [comment]
It is the "oh shit I broke develop" way. You'll be able to do the clean way later.

**Comment by nupplaphil** [comment]
Then I'll remove it immideatly ;-)

**Pull request merged**